stacks-common/src/util/vrf.rs

```diff
@@ -149,71 +114,76 @@ impl Default for VRFPrivateKey {
-                Ok(pk) => Some(VRFPrivateKey(pk)),
-                Err(_) => None,
-            },
-            Err(_) => None,
-        }
+        let bytes = hex_bytes(h).ok()?;
+        Self::from_bytes(bytes.as_slice())
     }
 
     pub fn from_bytes(b: &[u8]) -> Option<VRFPrivateKey> {
-        match ed25519_PrivateKey::from_bytes(b) {
-            Ok(pk) => Some(VRFPrivateKey(pk)),
-            Err(_) => None,
-        }
+        let signing_key = ed25519_dalek::SigningKey::try_from(b).ok()?;
+        Some(VRFPrivateKey(signing_key))
     }
 
     pub fn to_hex(&self) -> String {
         to_hex(self.as_bytes())
     }
+
+    pub fn as_bytes(&self) -> &[u8] {
+        self.0.as_bytes()
+    }
+
+    pub fn to_bytes(&self) -> [u8; 32] {
+        self.0.to_bytes()
+    }
 }
 
 impl VRFPublicKey {
-    pub fn from_private(pk: &VRFPrivateKey) -> VRFPublicKey {
-        VRFPublicKey(ed25519_PublicKey::from(&pk.0))
+    pub fn from_private(sk: &VRFPrivateKey) -> VRFPublicKey {
+        VRFPublicKey(sk.0.verifying_key())
     }
 
+    /// Verify that a given byte string is a well-formed EdDSA public
+    /// key (i.e. it's a compressed Edwards point that is valid), and return
+    /// a VRFPublicKey if so
     pub fn from_bytes(pubkey_bytes: &[u8]) -> Option<VRFPublicKey> {
-        match pubkey_bytes.len() {
-            32 => {
-                let mut pubkey_slice = [0; 32];
-                pubkey_slice.copy_from_slice(&pubkey_bytes[0..32]);
-
-                let checked_pubkey = CompressedEdwardsY(pubkey_slice);
-                match checked_pubkey.decompress() {
-                    Some(_) => {}
-                    None => {
-                        // invalid
-                        return None;
-                    }
-                }
-
-                match ed25519_PublicKey::from_bytes(&pubkey_slice) {
-                    Ok(key) => Some(VRFPublicKey(key)),
-                    Err(_) => None,
-                }
-            }
-            _ => None,
+        let pubkey_slice = pubkey_bytes.try_into().ok()?;
+
+        // NOTE: `ed25519_dalek::VerifyingKey::from_bytes` docs say
+        //  that this check must be performed by the caller, but as of
+        //  latest, it actually performs the check as well. However,
+        //  we do this check out of an abundance of caution because
+        //  that's what the docs say to do!
+
+        let checked_pubkey = CompressedEdwardsY(pubkey_slice);
+        if checked_pubkey.decompress().is_none() {
+            // invalid
+            return None;
         }
+
+        let key = ed25519_dalek::VerifyingKey::from_bytes(&pubkey_slice).ok()?;
+        Some(VRFPublicKey(key))
     }
 
     pub fn from_hex(h: &str) -> Option<VRFPublicKey> {
-        match hex_bytes(h) {
-            Ok(b) => VRF::check_public_key(&b),
-            Err(_) => None,
-        }
+        let bytes = hex_bytes(h).ok()?;
+        Self::from_bytes(bytes.as_slice())
```

Comment:
Slightly simpler:

```suggestion
        hex_bytes(h).ok().as_deref().and_then(Self::from_bytes)
```

Reply:
I'm a little conflicted about this kind of thing -- it's actually not obvious to me which of the above is simpler. Sometimes the functional chaining in `option` and `result` can be hard to decipher (especially because it often means that the `None` type survives to the end of the chain).

Reply:
I agree with @kantai , I find this kind of chaining to be less clear than the two line version.

Reply:
Am I the only one who finds chaining easier to read? It just seems like extra steps to take a value out of an `Option`/`Result` and then reconstruct another one, rather than using `map()`/`and_then()`

Reply:
For me, it depends on the situation. Sometimes I think it is much clearer to chain, other times less so. When there's multiple fallible steps, I think I usually prefer the non-chained version, because with chaining it can be hard to reason about how the error cases are being handled. Conversely, iteration is often much clearer to me when chained rather than manually looped (in particular things like finding an element or folding a sum).